File: samples/SpeechRecognition/src/MFCC.cpp
//
// Copyright © 2020 Arm Ltd and Contributors. All rights reserved.
// SPDX-License-Identifier: MIT
//

#include <cstdio>
#include <float.h>

#include "MFCC.hpp"
#include "MathUtils.hpp"


MfccParams::MfccParams(
        const float samplingFreq,
        const int numFbankBins,
        const float melLoFreq,
        const float melHiFreq,
        const int numMfccFeats,
        const int frameLen,
        const bool useHtkMethod,
        const int numMfccVectors):
        m_samplingFreq(samplingFreq),
        m_numFbankBins(numFbankBins),
        m_melLoFreq(melLoFreq),
        m_melHiFreq(melHiFreq),
        m_numMfccFeatures(numMfccFeats),
        m_frameLen(frameLen),
        m_numMfccVectors(numMfccVectors),

        /* Smallest power of 2 >= frame length. */
        m_frameLenPadded(pow(2, ceil((log(frameLen)/log(2))))),
        m_useHtkMethod(useHtkMethod)
{}

std::string MfccParams::Str()
{
    char strC[1024];
    snprintf(strC, sizeof(strC) - 1, "\n   \
            \n\t Sampling frequency:         %f\
            \n\t Number of filter banks:     %u\
            \n\t Mel frequency limit (low):  %f\
            \n\t Mel frequency limit (high): %f\
            \n\t Number of MFCC features:    %u\
            \n\t Frame length:               %u\
            \n\t Padded frame length:        %u\
            \n\t Using HTK for Mel scale:    %s\n",
             this->m_samplingFreq, this->m_numFbankBins, this->m_melLoFreq,
             this->m_melHiFreq, this->m_numMfccFeatures, this->m_frameLen,
             this->m_frameLenPadded, this->m_useHtkMethod ? "yes" : "no");
    return std::string{strC};
}

MFCC::MFCC(const MfccParams& params):
        _m_params(params),
        _m_filterBankInitialised(false)
{
    this->_m_buffer = std::vector<float>(
            this->_m_params.m_frameLenPadded, 0.0);
    this->_m_frame = std::vector<float>(
            this->_m_params.m_frameLenPadded, 0.0);
    this->_m_melEnergies = std::vector<float>(
            this->_m_params.m_numFbankBins, 0.0);

    this->_m_windowFunc = std::vector<float>(this->_m_params.m_frameLen);
    const float multiplier = 2 * M_PI / this->_m_params.m_frameLen;

    /* Create window function. */
    for (size_t i = 0; i < this->_m_params.m_frameLen; i++)
    {
        this->_m_windowFunc[i] = (0.5 - (0.5 * cos(static_cast<float>(i) * multiplier)));
    }
}

void MFCC::Init()
{
    this->_InitMelFilterBank();
}

float MFCC::MelScale(const float freq, const bool useHTKMethod)
{
    if (useHTKMethod)
    {
        return 1127.0f * logf (1.0f + freq / 700.0f);
    }
    else
    {
        /* Slaney formula for mel scale. */
        float mel = freq / freqStep;

        if (freq >= minLogHz)
        {
            mel = minLogMel + logf(freq / minLogHz) / logStep;
        }
        return mel;
    }
}

float MFCC::InverseMelScale(const float melFreq, const bool useHTKMethod)
{
    if (useHTKMethod)
    {
        return 700.0f * (expf (melFreq / 1127.0f) - 1.0f);
    }
    else
    {
        /* Slaney formula for mel scale. */
        float freq = freqStep * melFreq;

        if (melFreq >= minLogMel)
        {
            freq = minLogHz * expf(logStep * (melFreq - minLogMel));
        }
        return freq;
    }
}


bool MFCC::ApplyMelFilterBank(
        std::vector<float>&                 fftVec,
        std::vector<std::vector<float>>&    melFilterBank,
        std::vector<int32_t>&               filterBankFilterFirst,
        std::vector<int32_t>&               filterBankFilterLast,
        std::vector<float>&                 melEnergies)
{
    const size_t numBanks = melEnergies.size();

    if (numBanks != filterBankFilterFirst.size() ||
        numBanks != filterBankFilterLast.size())
    {
        printf("unexpected filter bank lengths\n");
        return false;
    }

    for (size_t bin = 0; bin < numBanks; ++bin)
    {
        auto filterBankIter = melFilterBank[bin].begin();
        float melEnergy = 1e-10; /* Avoid log of zero at later stages */
        const int32_t firstIndex = filterBankFilterFirst[bin];
        const int32_t lastIndex = filterBankFilterLast[bin];

        for (int32_t i = firstIndex; i <= lastIndex; ++i)
        {
            melEnergy += (*filterBankIter++ * fftVec[i]);
        }

        melEnergies[bin] = melEnergy;
    }

    return true;
}

void MFCC::ConvertToLogarithmicScale(std::vector<float>& melEnergies)
{
    float maxMelEnergy = -FLT_MAX;

    /* Container for natural logarithms of mel energies */
    std::vector <float> vecLogEnergies(melEnergies.size(), 0.f);

    /* Because we are taking natural logs, we need to multiply by log10(e).
     * Also, for wav2letter model, we scale our log10 values by 10 */
    constexpr float multiplier = 10.0 * /* default scalar */
                                 0.4342944819032518; /* log10f(std::exp(1.0))*/

    /* Take log of the whole vector */
    MathUtils::VecLogarithmF32(melEnergies, vecLogEnergies);

    /* Scale the log values and get the max */
    for (auto iterM = melEnergies.begin(), iterL = vecLogEnergies.begin();
         iterM != melEnergies.end(); ++iterM, ++iterL)
    {
        *iterM = *iterL * multiplier;

        /* Save the max mel energy. */
        if (*iterM > maxMelEnergy)
        {
            maxMelEnergy = *iterM;
        }
    }

    /* Clamp the mel energies */
    constexpr float maxDb = 80.0;
    const float clampLevelLowdB = maxMelEnergy - maxDb;
    for (auto iter = melEnergies.begin(); iter != melEnergies.end(); ++iter)
    {
        *iter = std::max(*iter, clampLevelLowdB);
    }
}

void MFCC::_ConvertToPowerSpectrum()
{
    const uint32_t halfDim = this->_m_params.m_frameLenPadded / 2;

    /* Handle this special case. */
    float firstEnergy = this->_m_buffer[0] * this->_m_buffer[0];
    float lastEnergy = this->_m_buffer[1] * this->_m_buffer[1];

    MathUtils::ComplexMagnitudeSquaredF32(
            this->_m_buffer.data(),
            this->_m_buffer.size(),
            this->_m_buffer.data(),
            this->_m_buffer.size()/2);

    this->_m_buffer[0] = firstEnergy;
    this->_m_buffer[halfDim] = lastEnergy;
}

std::vector<float> MFCC::CreateDCTMatrix(
        const int32_t inputLength,
        const int32_t coefficientCount)
{
    std::vector<float> dctMatix(inputLength * coefficientCount);

    /* Orthonormal normalization. */
    const float normalizerK0 = 2 * sqrt(1.0 / static_cast<float>(4*inputLength));
    const float normalizer = 2 * sqrt(1.0 / static_cast<float>(2*inputLength));

    const float angleIncr = M_PI/inputLength;
    float angle = angleIncr; /* we start using it at k = 1 loop */

    /* First row of DCT will use normalizer K0 */
    for (int32_t n = 0; n < inputLength; ++n)
    {
        dctMatix[n] = normalizerK0;
    }

    /* Second row (index = 1) onwards, we use standard normalizer */
    for (int32_t k = 1, m = inputLength; k < coefficientCount; ++k, m += inputLength)
    {
        for (int32_t n = 0; n < inputLength; ++n)
        {
            dctMatix[m+n] = normalizer *
                            cos((n + 0.5) * angle);
        }
        angle += angleIncr;
    }
    return dctMatix;
}

float MFCC::GetMelFilterBankNormaliser(
        const float&    leftMel,
        const float&    rightMel,
        const bool      useHTKMethod)
{
/* Slaney normalization for mel weights. */
    return (2.0f / (MFCC::InverseMelScale(rightMel, useHTKMethod) -
                    MFCC::InverseMelScale(leftMel, useHTKMethod)));
}

void MFCC::_InitMelFilterBank()
{
    if (!this->_IsMelFilterBankInited())
    {
        this->_m_melFilterBank = this->_CreateMelFilterBank();
        this->_m_dctMatrix = this->CreateDCTMatrix(
                this->_m_params.m_numFbankBins,
                this->_m_params.m_numMfccFeatures);
        this->_m_filterBankInitialised = true;
    }
}

bool MFCC::_IsMelFilterBankInited()
{
    return this->_m_filterBankInitialised;
}

void MFCC::_MfccComputePreFeature(const std::vector<float>& audioData)
{
    this->_InitMelFilterBank();

    /* TensorFlow way of normalizing .wav data to (-1, 1). */
    constexpr float normaliser = 1.0;
    for (size_t i = 0; i < this->_m_params.m_frameLen; i++)
    {
        this->_m_frame[i] = static_cast<float>(audioData[i]) * normaliser;
    }

    /* Apply window function to input frame. */
    for(size_t i = 0; i < this->_m_params.m_frameLen; i++)
    {
        this->_m_frame[i] *= this->_m_windowFunc[i];
    }

    /* Set remaining frame values to 0. */
    std::fill(this->_m_frame.begin() + this->_m_params.m_frameLen,this->_m_frame.end(), 0);

    /* Compute FFT. */
    MathUtils::FftF32(this->_m_frame, this->_m_buffer);

    /* Convert to power spectrum. */
    this->_ConvertToPowerSpectrum();

    /* Apply mel filterbanks. */
    if (!this->ApplyMelFilterBank(this->_m_buffer,
                                  this->_m_melFilterBank,
                                  this->_m_filterBankFilterFirst,
                                  this->_m_filterBankFilterLast,
                                  this->_m_melEnergies))
    {
        printf("Failed to apply MEL filter banks\n");
    }

    /* Convert to logarithmic scale */
    this->ConvertToLogarithmicScale(this->_m_melEnergies);
}

std::vector<float> MFCC::MfccCompute(const std::vector<float>& audioData)
{
    this->_MfccComputePreFeature(audioData);

    std::vector<float> mfccOut(this->_m_params.m_numMfccFeatures);

    float * ptrMel = this->_m_melEnergies.data();
    float * ptrDct = this->_m_dctMatrix.data();
    float * ptrMfcc = mfccOut.data();

    /* Take DCT. Uses matrix mul. */
    for (size_t i = 0, j = 0; i < mfccOut.size();
         ++i, j += this->_m_params.m_numFbankBins)
    {
        *ptrMfcc++ = MathUtils::DotProductF32(
                ptrDct + j,
                ptrMel,
                this->_m_params.m_numFbankBins);
    }

    return mfccOut;
}

std::vector<std::vector<float>> MFCC::_CreateMelFilterBank()
{
    size_t numFftBins = this->_m_params.m_frameLenPadded / 2;
    float fftBinWidth = static_cast<float>(this->_m_params.m_samplingFreq) / this->_m_params.m_frameLenPadded;

    float melLowFreq = MFCC::MelScale(this->_m_params.m_melLoFreq,
                                      this->_m_params.m_useHtkMethod);
    float melHighFreq = MFCC::MelScale(this->_m_params.m_melHiFreq,
                                       this->_m_params.m_useHtkMethod);
    float melFreqDelta = (melHighFreq - melLowFreq) / (this->_m_params.m_numFbankBins + 1);

    std::vector<float> thisBin = std::vector<float>(numFftBins);
    std::vector<std::vector<float>> melFilterBank(
            this->_m_params.m_numFbankBins);
    this->_m_filterBankFilterFirst =
            std::vector<int32_t>(this->_m_params.m_numFbankBins);
    this->_m_filterBankFilterLast =
            std::vector<int32_t>(this->_m_params.m_numFbankBins);

    for (size_t bin = 0; bin < this->_m_params.m_numFbankBins; bin++)
    {
        float leftMel = melLowFreq + bin * melFreqDelta;
        float centerMel = melLowFreq + (bin + 1) * melFreqDelta;
        float rightMel = melLowFreq + (bin + 2) * melFreqDelta;

        int32_t firstIndex = -1;
        int32_t lastIndex = -1;
        const float normaliser = this->GetMelFilterBankNormaliser(leftMel, rightMel, this->_m_params.m_useHtkMethod);

        for (size_t i = 0; i < numFftBins; i++)
        {
            float freq = (fftBinWidth * i); /* Center freq of this fft bin. */
            float mel = MFCC::MelScale(freq, this->_m_params.m_useHtkMethod);
            thisBin[i] = 0.0;

            if (mel > leftMel && mel < rightMel)
            {
                float weight;
                if (mel <= centerMel)
                {
                    weight = (mel - leftMel) / (centerMel - leftMel);
                }
                else
                {
                    weight = (rightMel - mel) / (rightMel - centerMel);
                }

                thisBin[i] = weight * normaliser;
                if (firstIndex == -1)
                {
                    firstIndex = i;
                }
                lastIndex = i;
            }
        }

        this->_m_filterBankFilterFirst[bin] = firstIndex;
        this->_m_filterBankFilterLast[bin] = lastIndex;

        /* Copy the part we care about. */
        for (int32_t i = firstIndex; i <= lastIndex; i++)
        {
            melFilterBank[bin].push_back(thisBin[i]);
        }
    }

    return melFilterBank;
}

File: python/pyarmnn/examples/common/mfcc.py
# Copyright © 2021 Arm Ltd and Contributors. All rights reserved.
# SPDX-License-Identifier: MIT

"""Class used to extract the Mel-frequency cepstral coefficients from a given audio frame."""

import numpy as np
import collections

MFCCParams = collections.namedtuple('MFCCParams', ['sampling_freq', 'num_fbank_bins', 'mel_lo_freq', 'mel_hi_freq',
                                                   'num_mfcc_feats', 'frame_len', 'use_htk_method', 'n_fft'])


class MFCC:

    def __init__(self, mfcc_params):
        self.mfcc_params = mfcc_params
        self.FREQ_STEP = 200.0 / 3
        self.MIN_LOG_HZ = 1000.0
        self.MIN_LOG_MEL = self.MIN_LOG_HZ / self.FREQ_STEP
        self.LOG_STEP = 1.8562979903656 / 27.0
        self._frame_len_padded = int(2 ** (np.ceil((np.log(self.mfcc_params.frame_len) / np.log(2.0)))))
        self._filter_bank_initialised = False
        self.__frame = np.zeros(self._frame_len_padded)
        self.__buffer = np.zeros(self._frame_len_padded)
        self._filter_bank_filter_first = np.zeros(self.mfcc_params.num_fbank_bins)
        self._filter_bank_filter_last = np.zeros(self.mfcc_params.num_fbank_bins)
        self.__mel_energies = np.zeros(self.mfcc_params.num_fbank_bins)
        self._dct_matrix = self.create_dct_matrix(self.mfcc_params.num_fbank_bins, self.mfcc_params.num_mfcc_feats)
        self.__mel_filter_bank = self.create_mel_filter_bank()
        self._np_mel_bank = np.zeros([self.mfcc_params.num_fbank_bins, int(self.mfcc_params.n_fft / 2) + 1])

        for i in range(self.mfcc_params.num_fbank_bins):
            k = 0
            for j in range(int(self._filter_bank_filter_first[i]), int(self._filter_bank_filter_last[i]) + 1):
                self._np_mel_bank[i, j] = self.__mel_filter_bank[i][k]
                k += 1

    def mel_scale(self, freq, use_htk_method):
        """
        Gets the mel scale for a particular sample frequency.

        Args:
            freq: The sampling frequency.
            use_htk_method: Boolean to set whether to use HTK method or not.

        Returns:
            the mel scale
        """
        if use_htk_method:
            return 1127.0 * np.log(1.0 + freq / 700.0)
        else:
            mel = freq / self.FREQ_STEP

        if freq >= self.MIN_LOG_HZ:
            mel = self.MIN_LOG_MEL + np.log(freq / self.MIN_LOG_HZ) / self.LOG_STEP
        return mel

    def inv_mel_scale(self, mel_freq, use_htk_method):
        """
        Gets the sample frequency for a particular mel.

        Args:
            mel_freq: The mel frequency.
            use_htk_method: Boolean to set whether to use HTK method or not.

        Returns:
            the sample frequency
        """
        if use_htk_method:
            return 700.0 * (np.exp(mel_freq / 1127.0) - 1.0)
        else:
            freq = self.FREQ_STEP * mel_freq

            if mel_freq >= self.MIN_LOG_MEL:
                freq = self.MIN_LOG_HZ * np.exp(self.LOG_STEP * (mel_freq - self.MIN_LOG_MEL))
            return freq

    def spectrum_calc(self, audio_data):
        return np.abs(np.fft.rfft(np.hanning(self.mfcc_params.frame_len + 1)[0:self.mfcc_params.frame_len] * audio_data,
                                  self.mfcc_params.n_fft))

    def log_mel(self, mel_energy):
        mel_energy += 1e-10  # Avoid division by zero
        return np.log(mel_energy)

    def mfcc_compute(self, audio_data):
        """
        Extracts the MFCC for a single frame.

        Args:
            audio_data: The audio data to process.

        Returns:
            the MFCC features
        """
        if len(audio_data) != self.mfcc_params.frame_len:
            raise ValueError(
                f"audio_data buffer size {len(audio_data)} does not match frame length {self.mfcc_params.frame_len}")

        audio_data = np.array(audio_data)
        spec = self.spectrum_calc(audio_data)
        mel_energy = np.dot(self._np_mel_bank.astype(np.float32),
                            np.transpose(spec).astype(np.float32))
        log_mel_energy = self.log_mel(mel_energy)
        mfcc_feats = np.dot(self._dct_matrix, log_mel_energy)
        return mfcc_feats

    def create_dct_matrix(self, num_fbank_bins, num_mfcc_feats):
        """
        Creates the Discrete Cosine Transform matrix to be used in the compute function.

        Args:
            num_fbank_bins: The number of filter bank bins
            num_mfcc_feats: the number of MFCC features

        Returns:
            the DCT matrix
        """

        dct_m = np.zeros(num_fbank_bins * num_mfcc_feats)
        for k in range(num_mfcc_feats):
            for n in range(num_fbank_bins):
                dct_m[(k * num_fbank_bins) + n] = (np.sqrt(2 / num_fbank_bins)) * np.cos(
                    (np.pi / num_fbank_bins) * (n + 0.5) * k)
        dct_m = np.reshape(dct_m, [self.mfcc_params.num_mfcc_feats, self.mfcc_params.num_fbank_bins])
        return dct_m

    def mel_norm(self, weight, right_mel, left_mel):
        """
        Placeholder function over-ridden in child class
        """
        return weight

    def create_mel_filter_bank(self):
        """
        Creates the Mel filter bank.

        Returns:
            the mel filter bank
        """
        # FFT calculations are greatly accelerated for frame lengths which are powers of 2
        # Frames are padded and FFT bin width/length calculated accordingly
        num_fft_bins = int(self._frame_len_padded / 2)
        fft_bin_width = self.mfcc_params.sampling_freq / self._frame_len_padded

        mel_low_freq = self.mel_scale(self.mfcc_params.mel_lo_freq, self.mfcc_params.use_htk_method)
        mel_high_freq = self.mel_scale(self.mfcc_params.mel_hi_freq, self.mfcc_params.use_htk_method)
        mel_freq_delta = (mel_high_freq - mel_low_freq) / (self.mfcc_params.num_fbank_bins + 1)

        this_bin = np.zeros(num_fft_bins)
        mel_fbank = [0] * self.mfcc_params.num_fbank_bins
        for bin_num in range(self.mfcc_params.num_fbank_bins):
            left_mel = mel_low_freq + bin_num * mel_freq_delta
            center_mel = mel_low_freq + (bin_num + 1) * mel_freq_delta
            right_mel = mel_low_freq + (bin_num + 2) * mel_freq_delta
            first_index = last_index = -1

            for i in range(num_fft_bins):
                freq = (fft_bin_width * i)
                mel = self.mel_scale(freq, self.mfcc_params.use_htk_method)
                this_bin[i] = 0.0

                if (mel > left_mel) and (mel < right_mel):
                    if mel <= center_mel:
                        weight = (mel - left_mel) / (center_mel - left_mel)
                    else:
                        weight = (right_mel - mel) / (right_mel - center_mel)

                    this_bin[i] = self.mel_norm(weight, right_mel, left_mel)

                    if first_index == -1:
                        first_index = i
                    last_index = i

            self._filter_bank_filter_first[bin_num] = first_index
            self._filter_bank_filter_last[bin_num] = last_index
            mel_fbank[bin_num] = np.zeros(last_index - first_index + 1)
            j = 0

            for i in range(first_index, last_index + 1):
                mel_fbank[bin_num][j] = this_bin[i]
                j += 1

        return mel_fbank


class AudioPreprocessor:

    def __init__(self, mfcc, model_input_size, stride):
        self.model_input_size = model_input_size
        self.stride = stride
        self._mfcc_calc = mfcc

    def _normalize(self, values):
        """
        Normalize values to mean 0 and std 1
        """
        ret_val = (values - np.mean(values)) / np.std(values)
        return ret_val

    def _get_features(self, features, mfcc_instance, audio_data):
        idx = 0
        while len(features) < self.model_input_size * mfcc_instance.mfcc_params.num_mfcc_feats:
            current_frame_feats = mfcc_instance.mfcc_compute(audio_data[idx:idx + int(mfcc_instance.mfcc_params.frame_len)])
            features.extend(current_frame_feats)
            idx += self.stride

    def mfcc_delta_calc(self, features):
        """
        Placeholder function over-ridden in child class
        """
        return features

    def extract_features(self, audio_data):
        """
        Extracts the MFCC features. Also calculates each features first and second order derivatives
        if the mfcc_delta_calc() function has been implemented by a child class.
        The matrix returned should be sized appropriately for input to the model, based
        on the model info specified in the MFCC instance.

        Args:
            audio_data: the audio data to be used for this calculation
        Returns:
            the derived MFCC feature vector, sized appropriately for inference
        """

        num_samples_per_inference = ((self.model_input_size - 1)
                                     * self.stride) + self._mfcc_calc.mfcc_params.frame_len

        if len(audio_data) < num_samples_per_inference:
            raise ValueError("audio_data size for feature extraction is smaller than "
                             "the expected number of samples needed for inference")

        features = []
        self._get_features(features, self._mfcc_calc, np.asarray(audio_data))
        features = np.reshape(np.array(features), (self.model_input_size, self._mfcc_calc.mfcc_params.num_mfcc_feats))
        features = self.mfcc_delta_calc(features)
        return np.float32(features)
